core/src/impl/Kokkos_GraphNodeThenImpl.hpp

```diff
@@ -23,28 +23,36 @@
 namespace Kokkos::Impl {
 
 // Helper for the 'then', such that the user can indeed pass a callable that
-// takes no argument.
-template <typename Functor>
+// takes no index.
+template <typename Functor, typename WorkTag>
 struct ThenWrapper {
   Functor functor;
-  template <typename T>
-  KOKKOS_FUNCTION void operator()(const T) const {
+
+  template <typename T, typename U = WorkTag>
+  KOKKOS_FUNCTION std::enable_if_t<std::is_void_v<U>> operator()(
+      const T) const {
     functor();
   }
+
+  template <typename T, typename U = WorkTag>
+  KOKKOS_FUNCTION std::enable_if_t<!std::is_void_v<U>> operator()(
+      const U, const T) const {
+    functor(U{});
+  }
```

Comment:
```suggestion
  template <typename T, typename U = WorkTag>
  KOKKOS_FUNCTION std::enable_if_t<!std::is_void_v<U>> operator()(
      const U& u, const T) const {
    functor(u);
  }
```
Don't know if there is a cost for constructing a variable of an empty type (which WorkTag should be)?

Reply:
I guess any decent compiler is able to optimize the work tag away if it's empty?

By the way, in the deprecated things, I see:
> Raise deprecation warnings if non-empty WorkTag class is used

And since the graph is experimental, I'm tempted to already forbid non-empty tags :angel:

Reply:
>And since the graph is experimental, I'm tempted to already forbid non-empty tags 👼
I didn't saw it was only deprecated, and actually enforced it for `single` so I won't blame you if you decide to do the same for `then` :).

Reply:
> I guess any decent compiler is able to optimize the work tag away if it's empty?

I thought a bit more about this, regardless of optimisation, `is_empty` doesn't preclude a type from having a constructor or a destructor with side effects ([see here for example](https://godbolt.org/z/PqbaY6bWT)) so we should avoid rebuilding WorkTags constantly.

Reply:
The worktag facility is intended to be used with tag classes.  The compiler explorer code snippet defines a default constructor with side effects and is not really relevant here.
The current implementation is fine.